### PR TITLE
gocollector: Added options to Go Collector for changing the

### PR DESCRIPTION
**Pull request by bwplotka**
Fixes https://github.com/prometheus/client_golang/issues/983

Also:

* Improved docs
* fixed TestMemStatsEquivalence, it was noop before (:
* Removed gc_cpu_fraction metric completely, since it's not working completely for Go1.17+

Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

### CODE REVIEW

**Comment by bwplotka** [comment]
cc @mknyszek - we would love your eyes on this  (:

File: prometheus/go_collector_latest_test.go [hunk 1]
```diff
@@ -227,9 +251,15 @@ func TestMemStatsEquivalence(t *testing.T) {
 	for i := 0; i < msRealValue.NumField(); i++ {
 		fr := msRealValue.Field(i)
 		ff := msFakeValue.Field(i)
-		switch typ.Kind() {
```
Comment (bwplotka) [inline prometheus/go_collector_latest_test.go:230]: Fun fact. This test was noop, because `typ.Kind` was always `struct` - we want field kind instead 🙈

**Comment by bwplotka** [comment]
After this is merged I will attempt to fix other issues:

https://github.com/prometheus/client_golang/issues/994
https://github.com/prometheus/client_golang/issues/995
https://github.com/prometheus/client_golang/issues/1026

I would also (in separate PR) propose to default to just `goRuntimeMemStats` for now instead of `goRuntimeMemStats | goRuntimeMetrics` set. I think migration to new metrics is important, but it has to be opt in for now IMO. We impacted a lot ppl by adding two digit more of metrics  and I am kind of missing the reason (users don't know how to leverage them yet) 🤔 Not sure, yet, let's discuss. (:

**Review by kakkoyun — COMMENTED** [review]
Overall looks good.

`prometheus/collectors/go_collector.go` file missing build flags to exclude versions older than <1.17, I guess, considering CI failures.

**Comment by bwplotka** [comment]
Fixed @kakkoyun 

@mknyszek I don't know much about `runtime/metrics.Read` and nothing about "stop-the-world" problem was written into its documentation (go docs). I assume we can remove section about the consequences of stop-the-world logic from Go Collector, right? Could you confirm? (:  I mean the following one:

```
// NewGoCollector returns a collector that exports metrics about the current Go
// process. This includes memory stats. To collect those, runtime.ReadMemStats
// is called. This requires to “stop the world”, which usually only happens for
// garbage collection (GC). Take the following implications into account when
// deciding whether to use the Go collector:
//
// 1. The performance impact of stopping the world is the more relevant the more
// frequently metrics are collected. However, with Go1.9 or later the
// stop-the-world time per metrics collection is very short (~25µs) so that the
// performance impact will only matter in rare cases. However, with older Go
// versions, the stop-the-world duration depends on the heap size and can be
// quite significant (~1.7 ms/GiB as per
// https://go-review.googlesource.com/c/go/+/34937).
//
// 2. During an ongoing GC, nothing else can stop the world. Therefore, if the
// metrics collection happens to coincide with GC, it will only complete after
// GC has finished. Usually, GC is fast enough to not cause problems. However,
// with a very large heap, GC might take multiple seconds, which is enough to
// cause scrape timeouts in common setups. To avoid this problem, the Go
// collector will use the memstats from a previous collection if
// runtime.ReadMemStats takes more than 1s. However, if there are no previously
// collected memstats, or their collection is more than 5m ago, the collection
// will block until runtime.ReadMemStats succeeds.
```

**Review by kakkoyun — APPROVED** [review]

File: prometheus/go_collector_latest_test.go (hunk 1 above)
Comment (mknyszek) [reply prometheus/go_collector_latest_test.go:230]: oops! good catch.

**Comment by mknyszek** [comment]
Hey sorry for the delay. If it matters anymore, LGTM! Thanks, and sorry I never got around to this.

**Pull request merged**